src/corehost/cli/bdl_processor.cpp

```diff
@@ -0,0 +1,316 @@
+	}
+}
+
+void bdl_processor_t::process_manifest_footer(int64_t& header_offset)
+{
+	seek(-manifest_footer_t::num_bytes_read(), SEEK_END);
+
+	manifest_footer_t* footer = manifest_footer_t::read(m_bundle);
+	header_offset = footer->header_offset;
+}
+
+void bdl_processor_t::process_manifest_header(int64_t header_offset)
+{
+	seek(header_offset, SEEK_SET);
+
+	manifest_header_t* header = manifest_header_t::read(m_bundle);
+
+	m_num_embedded_files = header->data.num_embedded_files;
+	m_bundle_id = header->bundle_id;
+}
+
+// Compute the final extraction location as:
+// m_extraction_dir = $DOTNET_BUNDLE_EXTRACT_BASE_DIR/<app>/<id>/...
+//
+// If DOTNET_BUNDLE_EXTRACT_BASE_DIR is not set in the environment, the 
+// base directory defaults to $TMPDIR/.net
+void bdl_processor_t::determine_extraction_dir()
+{
+	if (!pal::getenv(_X("DOTNET_BUNDLE_EXTRACT_BASE_DIR"), &m_extraction_dir))
+	{
+		if (!pal::get_temp_directory(m_extraction_dir))
+		{
+			trace::error(_X("Failed to get temp_dir"));
+			throw StatusCode::BundleExtractionFailure;
+		}
+
+		append_path(&m_extraction_dir, _X(".net"));
+	}
+
+	pal::string_t host_name = strip_executable_ext(get_filename(m_bundle_path));
+	append_path(&m_extraction_dir, host_name.c_str());
+	append_path(&m_extraction_dir, m_bundle_id.c_str());
+
+	trace::info(_X("Extraction Location [%s]"), m_extraction_dir.c_str());
+}
+
+// Compute the worker extraction location for this process, before the 
+// extracted files are committed to the final location
+// m_working_extraction_dir = $DOTNET_BUNDLE_EXTRACT_BASE_DIR/<app>/<proc-id-hex>
+void bdl_processor_t::create_working_extraction_dir()
+{
+	// Set the working extraction path
+	m_working_extraction_dir = get_directory(m_extraction_dir);
+	pal::char_t pid[32];
+	pal::snwprintf(pid, 32, _X("%x"), pal::get_pid());
+	append_path(&m_working_extraction_dir, pid);
+
+	create_directory_tree(m_working_extraction_dir);
+
+	trace::info(_X("Temporary Extraction Location [%s]"), m_working_extraction_dir.c_str());
+}
+
+// Create a file to be extracted out on disk, including any intermediate sub-directories.
+FILE* bdl_processor_t::create_extraction_file(const pal::string_t& relative_path)
+{
+	pal::string_t file_path = m_working_extraction_dir;
+	append_path(&file_path, relative_path.c_str());
+
+	// m_working_extraction_dir is assumed to exist, 
+	// so we only create sub-directories if relative_path contains directories
+	if (has_dirs_in_path(relative_path))
+	{
+		create_directory_tree(get_directory(file_path));
+	}
+
+	FILE* file = pal::file_open(file_path.c_str(), _X("wb"));
+
+	if (file == nullptr)
+	{
+		trace::error(_X("Failed to open file [%s] for writing"), file_path.c_str());
+		throw StatusCode::BundleExtractionIOError;
+	}
+
+	return file;
+}
+
+// Extract one file from the bundle to disk.
+void bdl_processor_t::extract_file(file_entry_t *entry)
+{
+	FILE* file = create_extraction_file(entry->relative_path);
+	uint8_t* buffer = new uint8_t[entry->data.size];
```

Comment:
The `data.size` is not validated in any way, so this may force us to allocate unreasonable amounts of memory.
Also, what if the embedded file is large (500MB, like a graphics, video and so on) - this will allocate the memory for the entire file, potentially causing swapping and so on.
We should stream the file chunk by chunk.